### DISCUSSION TITLE
Is there a simple way to convert a Bitmap to a Compose ImageBitmap?

### DISCUSSION THREAD

**Discussion by sproctor** [korlibs/korge#1246] [Q&A] [answered]
I'm trying to display a `Bitmap` in a Compose app.

I'm in the process of adding a JS target to a multiplatform library (https://github.com/sproctor/SignaturePad/). I currently have a function to convert a `Bitmap` to `ImageBItmap` for Android and JVM, but JS looks more difficult as there doesn't seem to be a simple interim format.

Do you know how to accomplish this? Or is there a simple way to display a `Bitmap` in a Compose app?

**Comment by soywiz** [comment]
Well, KorIM supports converting bitmaps to `Bitmap32` with `RGBA` pixels, so you can do whatever is needed.

In the case of JS I'm not sure, but it is possible to generate an URL for passing to a `<img>` with `HtmlNativeImage .toDataURL`:

https://github.com/korlibs/korge/blob/ccb4d13c5ac76b51e04c1272bd1e72b14c091a69/korim/src/jsMain/kotlin/com/soywiz/korim/format/NativeImageFormatProviderJs.kt#L118

You can also use Blob to generate shorter URLs:
https://github.com/korlibs/korge/blob/ccb4d13c5ac76b51e04c1272bd1e72b14c091a69/korim/src/jsMain/kotlin/com/soywiz/korim/format/NativeImageFormatProviderJs.kt#L187-L196

But you will need to revoke them later as in the snippet.

**Comment by sproctor** [comment]
Thanks. For my application, the image is being displayed using the `Image` function from Compose. I think the my best way forward is to write a Compose `ImageBitmap` wrapper for `Bitmap32`. That might perform better on every target, actually. When I have something working, I'll post it here.

**Reply by sproctor** [reply]
Nope, that wasn't a good idea. It's not supported on Android or JVM. Next try: writing a `Painter` for `Bitmap32`.

**Comment by sproctor** [comment, marked as answer]
Ok, writing a painter looked complicated. I opted for what appeared to be the easiest: convert a Korim `Bitmap` to a Skia `Bitmap`.

Here's the result:
```
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asComposeImageBitmap
import com.soywiz.korim.bitmap.Bitmap
import com.soywiz.korim.bitmap.Bitmap32
import org.jetbrains.skia.ColorAlphaType
import org.jetbrains.skia.ColorInfo
import org.jetbrains.skia.ColorType
import org.jetbrains.skia.ImageInfo

internal actual fun Bitmap.toComposeImageBitmap(): ImageBitmap {
    return toBMP32IfRequired().toSkiaBitmap().asComposeImageBitmap()
}

internal fun Bitmap32.toSkiaBitmap(): org.jetbrains.skia.Bitmap {
    val colorType = ColorType.N32
    val alphaType = if (premultiplied) ColorAlphaType.PREMUL else ColorAlphaType.OPAQUE
    val skiaColorSpace = null
    val colorInfo = ColorInfo(colorType, alphaType, skiaColorSpace)
    val imageInfo = ImageInfo(colorInfo, width, height)
    val bitmap = org.jetbrains.skia.Bitmap()
    val pixels = extractBytes()
    bitmap.allocPixels(imageInfo)
    bitmap.installPixels(pixels)
    return bitmap
}
```

**Reply by sproctor** [reply]
This works in my particular use case. I don't know if the alpha/premultiplied logic is correct, I don't know about the `ColorSpace` either.

**Comment by 7va** [comment]
I had the similar problem and better way that i found in Compose examples. [ByteArray -> ImageBitmap](https://github.com/JetBrains/compose-jb/blob/master/experimental/examples/imageviewer/shared/src/commonMain/kotlin/example/imageviewer/ImageBitmap.common.kt)
